File: python/algo/202507/algo3493.py
from typing import List

# 常用数据结构 - 7.1 并查集 - 基础

class UnionFind:
    def __init__(self, n:int):
        self.fa = list(range(n))
        self.cc = n                 # 连通块数目
    
    def find(self, x: int) -> int:
        if self.fa[x] != x:
            self.fa[x] = self.find(self.fa[x])
        return self.fa[x]

    def merge(self, u: int, v:int) -> None:
        x, y = self.find(u), self.find(v)
        if x != y:
            self.fa[x] = y
            self.cc -= 1

class Solution:
    def numberOfComponents(self, properties: List[List[int]], k: int) -> int:
        sets = list(map(set, properties))
        uf = UnionFind(len(properties))
        for i, a in enumerate(sets):
            for j, b in enumerate(sets[:i]):
                if len(a & b) >= k: uf.merge(j, i)

        return uf.cc

File: cplus/leetcode/202507/algo3493.cpp
#include <bits/stdc++.h>
using namespace std;

// 常用数据结构 - 7.1 并查集 - 基础
class UnionFind {                   // 并查集模板
    vector<int> fa;
public:
    int cc;         // 连通块个数
    UnionFind(int n): fa(n), cc(n) {
        ranges::iota(fa, 0);
    }

    int find(int x) {
        if (fa[x] != x) {
            fa[x] = find(fa[x]);
        }
        return fa[x];
    }

    void merge(int from, int to) {
        int x = find(from), y = find(to);
        if (x != y) {
            fa[x] = y; cc--;
        }
    }
};

class Solution {
public:
    int numberOfComponents(vector<vector<int>>& properties, int k) {
        // 参考题解，并查集
        int n = properties.size();
        vector<unordered_set<int>> sets(n);
        for (int i=0; i<n; i++) {
            sets[i] = unordered_set(properties[i].begin(), properties[i].end());
        }

        UnionFind uf(n);
        for (int i=0; i<n; ++i) for (int j=0; j<i; ++j) {
            int cnt = 0;
            for (int x:sets[j]){
                if (sets[i].contains(x)) cnt ++;
            }
            if (cnt >= k) uf.merge(i, j);
        }
        return uf.cc;
    }
};